data-processing/scripts/dataset_validator.py

```diff
@@ -0,0 +1,259 @@
+"""
+This CLI tool evaluate whether a datasets adhere to the expected schema.
+
+The tool takes two arguments:
+- dataset_folder: Path to the dataset folder. On UCloud this is danish-foundation-models/dfm-data/pre-training
+- datasheets_folder: Path to the datasheets folder. On GitHub this is 'danish-foundation-models/docs/datasheets'
+
+The tool will then check following attributes about the datasets in the dataset_folder:
+
+Datasheets:
+
+A datasheet of the same name should be located in the datasheets_folder and should contain the following information:
+- License
+- Languages
+
+Folder Structure:
+
+The dataset folder should contain one folder per dataset. Each dataset folder should have the following structure:
+
+```
+dataset_folder
+│
+└── dataset_name
+    │
+    ├── documents
+    │   ├── document1.jsonl.gz   # MANDATORY: one or more files containing the documents in the dataset
+    │   └── ...
+    │
+    └── attributes   # OPTIONAL: folder containing annotations from dataset cleaning
+```
+
+
+Schema:
+
+An entry in the dataset should adhere to the Document schema (defined below).
+
+```
+{
+    "id": "...",                      # MANDATORY: source-specific identifier
+    "text": "foo",                    # MANDATORY: textual content of the document
+    "source": "...",                  # MANDATORY: source of the data, such as peS2o, common-crawl, etc.
+    "added": "...",                   # MANDATORY: timestamp we acquired this data (time file was created), specified as
+                                        # YYYY-MM-DDTHH:MM:SS.TIMEZONE e.g 2021-04-13T12:52:46.000Z
+    "created": "..."                  # MANDATORY: timestamp when orig document was created (best-guess if not available),
+                                         # should be specified as a range;
+                                         # "YYYY-MM-DDTHH:MM:SS.TIMEZONE, YYYY-MM-DDTHH:MM:SS.TIMEZONE"
+    "metadata": {                     # OPTIONAL: source-specific metadata
+         ...
+     }
+}
+"""
+
+
+import argparse
+import gzip
+import json
+import logging
+from datetime import datetime
+from pathlib import Path
+from typing import Any
+
+import yaml
+from pydantic import BaseModel, field_validator
+from tqdm import tqdm
+
+logger = logging.getLogger(__name__)
+
+
+class Document(BaseModel):
+    id: str
+    text: str
+    source: str
+    added: str
+    created: str
+    metadata: dict[str, Any] = {}
+
+    @field_validator("added")
+    @classmethod
+    def check_timestamp_format(cls, v: str):
+        if not v:
+            raise ValueError("Timestamp 'added' is required.")
+        try:
+            datetime.strptime(v, "%Y-%m-%dT%H:%M:%S.%fZ")
+        except ValueError:
+            raise ValueError(
+                "Timestamp 'added' should be in the format 'YYYY-MM-DDTHH:MM:SS.TIMEZONE'.",
```

Comment:
I am fine with YYYY-MM-DD, Though I would prefer to keep it a range instead of allowing for both options.

Reply:
I am also fine with YYYY-MM-DD and I am also fine with YYYY-MM-DDTHH:MM:SS.SSSTIMEZONE or YYYY-MM-DDTHH:MM:SSTIMEZONE, but not YYYY-MM-DDTHH:MM:SS.TIMEZONE, because the last one is contradictory to the ISO standard formats.